include/ensmallen_bits/callbacks/timer_stop.hpp

```diff
@@ -0,0 +1,82 @@
+/**
+ * @file timer_stop.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of the timer stop callback function.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_CALLBACKS_TIMER_STOP_HPP
+#define ENSMALLEN_CALLBACKS_TIMER_STOP_HPP
+
+namespace ens {
+
+/**
+ * Timer stop function, based on the EndEpoch callback function.
+ */
+class TimerStop
+{
+ public:
+  /**
+   * Set up the print loss callback class with the width and output stream.
+   *
+   * @param time The timeout in seconds.
+   */
+  TimerStop(const double time) : time(time), reset(false)
+  { /* Nothing to do here. */ }
+
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void BeginEpoch(OptimizerType& /* optimizer */,
+                  FunctionType& /* function */,
+                  const MatType& /* coordinates */,
+                  const size_t /* epoch */,
+                  const double /* objective */)
+  {
+    // Start the timer.
+    if (!reset)
+    {
+      timer.tic();
+      reset = true;
+    }
+  }
+
+  /**
+   * Callback function called at the end of a pass over the data.
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param epoch The index of the current epoch.
+   * @param objective Objective value of the current point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EndEpoch(OptimizerType& optimizer,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const size_t /* epoch */,
+                const double /* objective */)
+  {
+    if (timer.toc() > time)
+    {
+      Info << "Timer timeout reached; terminate optimization." << std::endl;
+      optimizer->Terminate() = true;
```

Comment:
Another idea would be to return a `bool` instead indicating whether optimization should continue.  That might allow us to avoid adding a `terminate` field to every optimizer.

Reply:
True, this replicated keras and I guess also is shorted since we don't have to write ``terminate |= Callback(...);`` for each callback. But I can see the benefit, to avoid the extra ``Terminate`` function, so not sure if you like to go with the ``bool`` I make the change.

Reply:
Yeah, personally I think the `bool` approach is better here, since it doesn't require us to add a member to every optimizer.